plonky2/src/fri/batch_prover.rs

```diff
@@ -0,0 +1,463 @@
+#[cfg(not(feature = "std"))]
+use alloc::vec;
+#[cfg(not(feature = "std"))]
+use alloc::vec::Vec;
+
+use plonky2_field::extension::flatten;
+use plonky2_field::types::Field;
+use plonky2_maybe_rayon::*;
+use plonky2_util::reverse_index_bits_in_place;
+
+use crate::field::extension::{unflatten, Extendable};
+use crate::field::polynomial::{PolynomialCoeffs, PolynomialValues};
+use crate::fri::proof::{FriInitialTreeProof, FriProof, FriQueryRound, FriQueryStep};
+use crate::fri::prover::{fri_proof_of_work, FriCommitedTrees};
+use crate::fri::FriParams;
+use crate::hash::field_merkle_tree::FieldMerkleTree;
+use crate::hash::hash_types::RichField;
+use crate::hash::merkle_tree::MerkleTree;
+use crate::iop::challenger::Challenger;
+use crate::plonk::config::GenericConfig;
+use crate::plonk::plonk_common::reduce_with_powers;
+use crate::timed;
+use crate::util::timing::TimingTree;
+
+/// Builds a batch FRI proof.
+pub fn batch_fri_proof<F: RichField + Extendable<D>, C: GenericConfig<D, F = F>, const D: usize>(
+    initial_merkle_trees: &[&FieldMerkleTree<F, C::Hasher>],
+    lde_polynomial_coeffs: PolynomialCoeffs<F::Extension>,
+    lde_polynomial_values: &mut [PolynomialValues<F::Extension>],
+    challenger: &mut Challenger<F, C::Hasher>,
+    fri_params: &FriParams,
+    timing: &mut TimingTree,
+) -> FriProof<F, C::Hasher, D> {
+    let n = lde_polynomial_coeffs.len();
+    assert_eq!(lde_polynomial_values[0].len(), n);
+    // The polynomial vectors should be sorted by degree, from largest to smallest, with no duplicate degrees.
```

Comment:
Could we just do the sorting in here, instead of taking it as a pre-condition?

(If the input is already sorted, then sorting only takes O(n) comparisons in Rust, the same as your check.)

Reply:
We cannot change their orders here.  We also need to change their corresponding FRI instances.

Reply:
Thanks.  Yes, I figured that one out by myself just now while playing with the code.

Oh, well, at least it helped me understand what's going on better. :)